x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmSingleNodeTests.java

```diff
@@ -134,6 +151,75 @@ protected boolean addMockHttpTransport() {
         return false;
     }
 
+    public void testGrantApiKeyForJWT() throws Exception {
+        final JWTClaimsSet.Builder jwtClaims = new JWTClaimsSet.Builder();
+        final String subject;
+        final String sharedSecret;
+        // id_token or access_token
+        if (randomBoolean()) {
+            subject = "me";
+            // JWT "id_token" valid for jwt0
+            jwtClaims.audience("es-01")
+                .issuer("my-issuer-01")
+                .subject(subject)
+                .claim("groups", "admin")
+                .issueTime(Date.from(Instant.now()))
+                .expirationTime(Date.from(Instant.now().plusSeconds(600)))
+                .build();
+            sharedSecret = jwt0SharedSecret;
+        } else {
+            subject = "me@example.com";
+            // JWT "access_token" valid for jwt2
+            jwtClaims.audience("es-03")
+                .issuer("my-issuer-03")
+                .subject("user-03")
+                .claim("groups", "admin")
+                .claim("email", subject)
+                .issueTime(Date.from(Instant.now()))
+                .expirationTime(Date.from(Instant.now().plusSeconds(300)));
+            sharedSecret = jwt2SharedSecret;
+        }
+        {
+            // JWT is valid but the client authentication is NOT
+            GrantApiKeyRequest grantApiKeyRequest = getGrantApiKeyForJWT(getSignedJWT(jwtClaims.build()), randomFrom("WRONG", null));
+            ElasticsearchSecurityException e = expectThrows(
+                ElasticsearchSecurityException.class,
+                () -> client().execute(GrantApiKeyAction.INSTANCE, grantApiKeyRequest).actionGet()
+            );
+            assertThat(e.getMessage(), containsString("unable to authenticate user"));
```

Comment:
nit: is there anything more specific to tell that the failure was due to the shared secret ? (if not should there be?)

Reply:
No, not right, not in the exception we throw. The "mismatched shared secret" cause IS logged though.

If the shared secret is wrong, ultimately no realm is going to be able to authenticate the token so authn gets to this point https://github.com/elastic/elasticsearch/blob/e03b0a5329f7cb230831dd566724956b6c4e855e/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RealmsAuthenticator.java#L256. As you can see we only log the authn result messages, but the error is always the same.
That's because, In the past, there was controversy over the details that we show to unauthenticated users (security scanners complain about it).

We could theoretically show those details in the response to the grant API Key API though, but it's not an easy code change because we reuse so much of the code for regular authn (which is designed to hide those details).